src/main/java/org/broadinstitute/hellbender/tools/walkers/rnaseq/GeneExpressionEvaluation.java

```diff
@@ -0,0 +1,667 @@
+                throw new UserException.CouldNotCreateOutputFile(file, " is not writable");
+            }
+        } else {
+            if (!Files.isWritable(file.getParentFile().toPath())) {
+                throw new UserException.CouldNotCreateOutputFile(file, " is not writable");
+            }
+        }
+    }
+
+    private Gff3BaseData shrinkBaseData(final Gff3BaseData baseData) {
+        //remove all but featureLabelKey attributes
+        final Map<String, List<String>> shrunkAttributes = baseData.getAttributes().entrySet().stream().filter(e -> e.getKey().equals(featureLabelKey.getKey())).collect(Collectors.toMap(Map.Entry::getKey,Map.Entry::getValue));
+        return new Gff3BaseData(baseData.getContig(), baseData.getSource(), baseData.getType(), baseData.getStart(), baseData.getEnd(), baseData.getScore(), baseData.getStrand(), baseData.getPhase(), shrunkAttributes);
+    }
+
+    private void addGroupingFeature(final Gff3BaseData groupingBaseData, final List<Interval> overlappingFeatures) {
+        final String geneLabel = featureLabelKey.getValue(groupingBaseData);
+        if (geneLabel == null) {
+            throw new UserException("no geneid field " + featureLabelKey + " found in feature at " + groupingBaseData.getContig() + ":" + groupingBaseData.getStart() + "-" + groupingBaseData.getEnd());
+        }
+
+        featureCounts.put(groupingBaseData, new Coverage(0, 0));
+        for (final Interval overlappingFeature : overlappingFeatures) {
+            featureOverlapDetector.addLhs(Pair.of(groupingBaseData, overlappingFeature), overlappingFeature);
+        }
+    }
+
+
+    static boolean inGoodPair(final GATKRead read, int minimumMappingQuality) {
+
+        boolean ret = !read.mateIsUnmapped() && read.isProperlyPaired() && read.getContig().equals(read.getMateContig()) &&
+                read.isReverseStrand() != read.mateIsReverseStrand();
+
+        if (ret) {
+            if (!read.hasAttribute(SAMTag.MQ.toString())) {
+                throw new GATKException("Mate quality must be included.  Consider running FixMateInformation.");
+            }
+            ret = ret && read.getAttributeAsInteger(SAMTag.MQ.toString()) >= minimumMappingQuality;
+        }
+
+        if (ret) {
+            if (read.isReverseStrand()) {
+                ret = ret && read.getEnd() >= read.getMateStart();
+            } else {
+                if (!read.hasAttribute(SAMTag.MC.toString())) {
+                    throw new GATKException("Mate cigar must be present.  Consider running FixMateInformation.");
+                }
+                final Cigar mateCigar = TextCigarCodec.decode(read.getAttributeAsString(SAMTag.MC.toString()));
+                ret = ret && read.getStart() <= read.getMateStart() + mateCigar.getReferenceLength();
+            }
+        }
+        return ret;
+    }
+
+    static List<Interval> getAlignmentIntervals(final GATKRead read, final boolean spliced, final int minimumMappingQuality) {
```

Comment:
I think it actually could be called on an unpaired read, and in fact is called on reads which are not in good pairs.  All the logic which relies on a mate sits inside an `inGoodPair` check.

Reply:
I think it would be helpful to have a comment here that if a read is paired it should only be called on one read (otherwise you'd be double counting).

Reply:
done